### ISSUE TITLE
NumPy or Python crashing with segfault (migrated from Trac #378)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#985]
Original ticket http://projects.scipy.org/numpy/ticket/378
Reported 2006-11-11 by trac user tecki, assigned to unknown.

I just ran the following little python script:

```
import numpy

a = numpy.arange(100 * 100)
a = a.reshape(100, 100)
b = a

def p(x, y):
        print x, y
        return x

v = numpy.vectorize(p)
print v(a,b)
```

and I god a segmentation fault. I considered this strange, since there should
be no segfaults using python, but exceptions, no?

OK, I am using slackware linux on an i386, tried their packages python-2.4.1, python-2.4.3 and a self-compiled python-2.5, all with numpy-1.0. Since the
python version doesn't matter, I guessed it is the numpy package which makes problems.

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2006-11-11

I can't reproduce this behaviour using r3430.

**Comment by thouis** [comment]
Comment in Trac by atmention:alberts, 2006-11-12

See if you can get a stack trace using GDB.

**Comment by thouis** [comment]
Comment in Trac by trac user tecki, 2006-11-12

OK, I tried to run with r3435 (at least print numpy.version.version says
1.0.1.dev3435), same segfault. Stack trace by GDB:

```
#0  0x0807564e in PyInt_AsLong (op=0x81720f4) at Objects/intobject.c:155
155             if (op == NULL || (nb = op->ob_type->tp_as_number) == NULL ||
(gdb) bt
#0  0x0807564e in PyInt_AsLong (op=0x81720f4) at Objects/intobject.c:155
#1  0xb7c19be9 in LONG_setitem (op=0x81720f4, ov=0x82ab548 "", ap=0x823ac40)
    at arraytypes.inc.src:133
#2  0xb7c21732 in OBJECT_to_LONG (ip=0x82a1900, op=0x82ab548, n=10000,
    aip=0x82829f0, aop=0x823ac40) at arraytypes.inc.src:784
#3  0xb7c48a82 in PyArray_CastTo (out=0x823ac40, mp=0x82829f0)
    at arrayobject.c:7607
#4  0xb7c5047e in PyArray_CastToType (mp=0x82829f0, at=0xb7c78160, fortran_=0)
    at arrayobject.c:7380
#5  0xb7c5956a in array_cast (self=0x82829f0, args=0xb7b3e30c)
    at arraymethods.c:661
#6  0x080c7b4a in PyEval_EvalFrameEx (f=0x8289e8c, throwflag=0)
    at Python/ceval.c:3566
#7  0x080c8c02 in PyEval_EvalCodeEx (co=0x8289e8c, globals=0x1,
    locals=0x8172100, args=0xb7d4002c, argcount=3, kws=0x0, kwcount=0,
    defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:2833
#8  0x08113c12 in function_call (func=0xb79840d4, arg=0xb7979bbc, kw=0x0)
    at Objects/funcobject.c:517
#9  0x0805e47e in PyObject_Call (func=0x823ac40, arg=0xb7979bbc, kw=0x0)
    at Objects/abstract.c:1860
#10 0x08067777 in instancemethod_call (func=0x1, arg=0xb7979bbc, kw=0x0)
    at Objects/classobject.c:2493
#11 0x0805e47e in PyObject_Call (func=0x823ac40, arg=0xb7b306cc, kw=0x0)
    at Objects/abstract.c:1860
#12 0x0809f4ee in slot_tp_call (self=0xb7b3ed6c, args=0xb7b306cc, kwds=0x0)
    at Objects/typeobject.c:4590
#13 0x0805e47e in PyObject_Call (func=0x823ac40, arg=0xb7b306cc, kw=0x0)
    at Objects/abstract.c:1860
#14 0x080c2d12 in PyEval_EvalFrameEx (f=0x8177e2c, throwflag=0)
    at Python/ceval.c:3777
#15 0x080c8c02 in PyEval_EvalCodeEx (co=0x8177e2c, globals=0x1,
    locals=0x8172100, args=0xb7d4002c, argcount=0, kws=0x0, kwcount=0,
    defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:2833
#16 0x080c8d65 in PyEval_EvalCode (co=0xb7d06650, globals=0xb7d58a44,
    locals=0xb7d58a44) at Python/ceval.c:494
#17 0x080e82c6 in PyRun_FileExFlags (fp=0x816e008,
    filename=0xbfc7e872 "hilf.py", start=257, globals=0xb7d58a44,
    locals=0xb7d58a44, closeit=1, flags=0xbfc7da7c) at Python/pythonrun.c:1264
#18 0x080e9090 in PyRun_SimpleFileExFlags (fp=0x816e008,
    filename=0xbfc7e872 "hilf.py", closeit=1, flags=0xbfc7da7c)
    at Python/pythonrun.c:870
#19 0x08056bef in Py_Main (argc=1, argv=0xbfc7dbb4) at Modules/main.c:496
#20 0x08056209 in main (argc=2, argv=0xbfc7dbb4) at ./Modules/python.c:23
```

I also tried to run the same script on windows (using numpy-1.0),
with the result of a general protection fault. I'll try to compile with 
another compiler, I've been using GCC 3.4.6, which version are you using?

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2006-11-12

I ran the code through valgrind under numpy r3435, and I still see no memory errors on my platform.

**Comment by thouis** [comment]
Comment in Trac by trac user tecki, 2006-11-13

I think I found the bug. In ''ufuncobject.c'' line 1975, there is a
''Py_XDECREF''. This decreases the reference counter of the object
in the buffer ''castbuf''. When this buffer is re-used in the next
pass in the loop, in line 1906 the refcounter gets decreased again.
Changing the ''Py_XDECREF'' to ''Py_CLEAR'' solved my problem.
(I still get spurious segfaults, but at different places). 

I think this does not introduce new bugs, since the content of the
buffer should not be used anymore anyways once ''Py_XDECREF''ed.

So I propose the following patch:

```
--- ufuncobject.c       (révision 3435)
+++ ufuncobject.c       (copie de travail)
atmention:atmention: -1972,7 +1972,7 atmention:atmention:
                                                                size = loop->leftover;

                                                        for (j=0; j<size; j++) {
-                                                               Py_XDECREF(*objptr);
+                                                               Py_CLEAR(*objptr);
                                                                objptr += 1;
                                                        }
                                                }
```

**Comment by thouis** [comment]
Comment in Trac by atmention:nilswagner01, 2006-11-14

I can reproduce the segfault.
Numpy version 1.0.1.dev3435
    [[   0    1    2 ...,   97   98   99]
     [9900 9901 9902 ..., 9997 9998 9999]
     [9900 9901 9902 ..., 9997 9998 9999]
     ...,
     [9900 9901 9902 ..., 9997 9998 9999]
     [9900 9901 9902 ..., 9997 9998 9999]
     [9900 9901 9902 ..., 9997 9998 9999]]

```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 46912509653888 (LWP 25266)]
0x00002aaaaac77b8d in Py_Main () from /usr/lib64/libpython2.4.so.1.0


(gdb)bt
...
#2354 0x0000000000000000 in ?? ()
#2355 0x0000000000000000 in ?? ()
#2356 0x0000000000000000 in ?? ()
#2357 0x0000000000000000 in ?? ()

#2358 0x0000000000000000 in ?? ()
#2359 0x0000000000000000 in ?? ()
Cannot access memory at address 0x7fffffb16000
```

**Comment by thouis** [comment]
Comment in Trac by trac user tecki, 2006-11-14

nils, can you try if this segfault disappears after applying the patch mentioned above, and could someone explain me, given the patch turns out to work (it does
for me) how this patch can make its way into the sources?

**Comment by thouis** [comment]
Comment in Trac by atmention:nilswagner01, 2006-11-14

I would prefer an official patch.

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2006-11-14

There's nothing wrong with using contributed patches like the one above -- someone who is intricately familiar with the code just has to review it first, and unfortunately Travis (who probably knows this code best) is currently on numpy leave.

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2006-11-14

I don't have much time to look at this, right now, but the reference counting going on was carefully implemented several times previously.  There may still be a problem, but I don't think it is as simple as getting rid of a DECREF.   All of the loops (casting and the function call itself) also affect the reference count of the objects pointed to. 

I'll need a better explanation with more careful tracking of the reference counts before accepting the patch.

**Comment by thouis** [comment]
Comment in Trac by trac user tecki, 2006-11-15

Well, I am ''not'' just getting rid of a DECREF. Instead I am changing a DECREF to
a CLEAR. This CLEAR does exactly the same DECREF, only that it sets the pointer
to NULL afterwards. This prevents a second unwanted DECREF in the next loop, as described above. One could argue that we could just let the DECREF be done in the
next pass of the loop: but that would be insane, since the last time we go through
the loop, the buffer will ''not'' be reused afterwards, we would have to clean it
in a step later.

I actually did a very careful tracking of references, saw that the references
of objects like int(158) (the 158 is an arbitrary example) go down to 0 and even
negative (that should ''really'' not happen). With my patch the refcounts go exactly
to the right values. I'll write something that illustrates that.

**Comment by thouis** [comment]
Comment in Trac by trac user tecki, 2006-11-15

I just wrote the mentioned script:

```
import numpy
import sys

a = numpy.arange(100 * 100)
b = a.reshape(100, 100)
c = b

# create enough references to all the values
# (otherwise script crashes on my system)
l = [a.tolist() for i in range(30)]

i = 1

def p(x, y):
        return 123

v = numpy.vectorize(p)

for j in range(15):
        d = v(b, c)
        print sys.getrefcount(i)
```

With the unpatched version, this gives:

```
548
547
544
542
540
538
536
534
532
530
528
526
524
522
520
```

so, the refcounter gets decreased by 2 everytime the vectorized function is called.
On the other hand, my patched version gives:

```
695
696
695
695
695
695
695
695
695
695
695
695
695
695
695
```

So no changes each round. (The different value from the beginning is due to the
fact that the unpatched and the patched python versions are actually different
pythons, for the former 2.4.3, for the latter 2.5)

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2006-11-17

Thanks for this detective work.  Getting the reference counting of OBJECT arrays right in the buffered section of the ufunc loop is certainly challenging.  I thought I had accounted for the cases.  Look in numpy/numpy/doc/ufunc.txt for some of the explanation. 

At any rate.  Using Py_CLEAR certaintly doesn't hurt as the object in the casting buffer should not
be re-used anyway and it seems to help.  Perhaps what is happening, is that when the buffer is re-used it is not filling up and a new decref is being performed on an "un-copied" object. 

So, I've accepted and applied the changes.  I changed both XDECREF's in that section of code to CLEAR. 

Thanks very much,

Travis
